### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by blond**
Get rid of `BemNaming` class.

### Why?

* To create namespace with `BemNaming` class there was necessity to copy methods (with `bind`).
* Initialization of user options was not obvious.

### What was done?

* Added `createNaming`, `init` and `buildRegex` functions.
* Added `origin` preset to initialize options with default values.
* Throw error if preset is unknown.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -15,259 +15,303 @@ var TYPES = {
...
+ * @readonly
+ */
+var WORD_PATTERN = '[a-zA-Z0-9]+(?:-[a-zA-Z0-9]+)*';
+
+/**
+ * Presets of options for various naming.
+ * @readonly
+ */
+var presets = {
+    'origin': {
```
Comment (tadatuta) [inline index.js]: I'd keep it without quotes

File: index.js [hunk 2]
```diff
@@ -15,259 +15,303 @@ var TYPES = {
...
+        delims: {
+            elem: '__',
+            mod: { name: '--', val: '_' }
+        },
+        wordPattern: WORD_PATTERN
+    }
+};
+
+/**
+ * It is necessary to not create new instances for the same custom naming.
```
Comment (tadatuta) [inline index.js]: to not -> not to

File: index.js [hunk 3]
```diff
@@ -15,259 +15,303 @@ var TYPES = {
...
  *
- * @param {Object} options              Options.
- * @param {String} options.elem         Separates element's name from block.
- * @param {String} options.mod          Separates names and values of modifiers from blocks and elements.
- * @param {String} options.modVal       Separates separate value of modifier from name of modifier.
- * @param {String} options.wordPattern  Defines which symbols can be used for block, element and modifier's names.
- * @name BemNaming
- * @class
+ * @param {Object} [options]              Options.
+ * @param {String} [options.elem=__]      Separates element's name from block.
```
Comment (tadatuta) [inline index.js]: shouldn't it be `[options.elem='__']` (with quotes)?

File: index.js [hunk 4]
```diff
@@ -15,259 +15,303 @@ var TYPES = {
...
+ * @param {String|Object} [options.mod=_] Separates modifiers from blocks and elements.
+ * @param {String} [options.mod.name=_]   Separates separate name of modifier from blocks and elements.
```
Comment (tadatuta) [inline index.js]: Separates separate

File: index.js [hunk 5]
```diff
@@ -15,259 +15,303 @@ var TYPES = {
...
+     * @returns {Boolean}
      */
-    this.modDelim = options.modName;
+    function validate(str) {
+        return regex.test(str);
+    }
+
     /**
-     * String to separate value of modifier from name of modifier.
+     * Parses string into BEM-naming.
```
Comment (tadatuta) [inline index.js]: what is `BEM-naming`?

File: index.js [hunk 6]
```diff
@@ -15,259 +15,303 @@ var TYPES = {
...
+        return res;
+    }
 
-    var notation = {
-            block: executed[1] || executed[4]
-        },
-        elem = executed[5],
-        modName = executed[2] || executed[6];
+    /**
+     * Returns a string indicating the type of the BEM entity.
```
Comment (tadatuta) [inline index.js]: Returns a string indicating type of a BEM entity

File: index.js [hunk 7]
```diff
@@ -15,259 +15,303 @@ var TYPES = {
...
+function init(options) {
+    options || (options = {});
+
+    if (typeof options === 'string') {
+        var preset = presets[options];
 
-        if (modVal && modVal !== true) {
-            res += this.modValDelim + modVal;
+        if (!preset) {
+            throw new Error('The `' + options + '` naming is unknown!');
```
Comment (tadatuta) [inline index.js]: !!1111

File: index.js (hunk 3 above)
Comment (blond) [inline index.js]: why?

**Comment by blond** [comment]
:up:

**Comment by blond** [comment]
:up:

**Pull request merged**